db/db_impl_secondary.cc

```diff
@@ -52,12 +58,195 @@ Status DBImplSecondary::Recover(
+        log_number, std::unique_ptr<log::Reader::Reporter>(reporter)));
+    log_reader_status_.insert(
+        std::make_pair(log_number, std::unique_ptr<Status>(status)));
+  }
+  iter = log_readers_.find(log_number);
+  assert(iter != log_readers_.end());
+  *log_reader = iter->second.get();
+  return Status::OK();
+}
+
+// After manifest recovery, replay WALs and refresh log_readers_ if necessary
+// REQUIRES: log_numbers are sorted in ascending order
+Status DBImplSecondary::RecoverLogFiles(
+    const std::vector<uint64_t>& log_numbers, SequenceNumber* next_sequence,
+    bool /*read_only*/) {
+  mutex_.AssertHeld();
+  Status status;
+  for (auto log_number : log_numbers) {
+    log::FragmentBufferedReader* reader = nullptr;
+    status = MaybeInitLogReader(log_number, &reader);
+    if (!status.ok()) {
+      return status;
+    }
+    if (reader == nullptr) {
+      return Status::NotFound();
+    }
+  }
+  for (auto log_number : log_numbers) {
+    log::FragmentBufferedReader* reader = log_readers_.at(log_number).get();
+    assert(reader != nullptr);
+    // Manually update the file number allocation counter in VersionSet.
+    versions_->MarkFileNumberUsed(log_number);
+
+    // Determine if we should tolerate incomplete records at the tail end of the
+    // Read all the records and add to a memtable
+    std::string scratch;
+    Slice record;
+    WriteBatch batch;
+
+    while (reader->ReadRecord(&record, &scratch,
+                              immutable_db_options_.wal_recovery_mode) &&
+           status.ok()) {
+      if (record.size() < WriteBatchInternal::kHeader) {
+        reader->GetReporter()->Corruption(
+            record.size(), Status::Corruption("log record too small"));
+        continue;
+      }
+      WriteBatchInternal::SetContents(&batch, record);
+      // do not check sequence number because user may toggle disableWAL
+      // between writes which breaks sequence number continuity guarantee
+
+      // If column family was not found, it might mean that the WAL write
+      // batch references to the column family that was dropped after the
+      // insert. We don't want to fail the whole write batch in that case --
+      // we just ignore the update.
+      // That's why we set ignore missing column families to true
+      bool has_valid_writes = false;
+      status = WriteBatchInternal::InsertIntoWithoutSchedulingFlush(
+          &batch, column_family_memtables_.get(), nullptr, true, log_number,
+          this, false /* concurrent_memtable_writes */, next_sequence,
+          &has_valid_writes, seq_per_batch_, batch_per_txn_);
+      if (!status.ok()) {
+        // We are treating this as a failure while reading since we read valid
+        // blocks that do not form coherent data
+        reader->GetReporter()->Corruption(record.size(), status);
+        continue;
+      }
+    }
+
+    if (!status.ok()) {
+      return status;
+    }
+
+    auto last_sequence = *next_sequence - 1;
```

Comment:
After successfully finishing one WAL, we should remove it from the map. The same holds for log repoter and status.

Reply:
I guess we can use the return value of `ReadRecord` to determine whether a WAL is finished? but if we do that wouldn't `log_reader_` always be empty after `RecoverLogFiles` finishes?

Reply:
We cannot delete the one with largest log number.

Reply:
Now that we need to insert/remove dynamically, how would you feel about using a struct for log reader, log reporter, log status? Therefore we have a map from log number to this struct.
I know I did not do this for manifest reader, but it's because once there is a new MANIFEST, we can stop reading the previous one and read the new one. Therefore no map is needed.

Reply:
I can add a struct to hold reader etc and make a single map <log number, struct unique_ptr>. 
At the end of `RecoverLogFiles` remove all elements from the map except the one with largest log number, how does this sound?

Reply:
Sounds good!

Reply:
After processing a WAL which is not the last WAL, we can remove the corresponding element from the map, right?